### PR TITLE
[PVR] fix timer type display for 'invalid' types

### PR DESCRIPTION
**Pull request by metaron-uk**
Allows 'invalid' timer types to be correctly displayed in the Timer Settings dialog
(See forum http://forum.kodi.tv/showthread.php?tid=307538)

## Description
Currently a timer type can be 'skipped' when building the list of valid timer types to present in the Timer Settings dialog. The result is that not all timer types are displayed properly.
This change adds the current type to the list if it would otherwise be skipped so the settings can be displayed correctly.

## Motivation and Context
Ensures timers are always displayed correctly.

## How Has This Been Tested?
On x86 (linux) with pvr.mythtv client.

## Screenshots:
Display of an 'Unhandled' timer type
screenshot062
After (How this type should be presented):
screenshot063
screenshot064
Before (How this type was previously presented):
screenshot065
screenshot066

## Types of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by metaron-uk** [comment]
@ksooo - if you like this, let me know and I'll create a backport PR as I think it ought to go in the next Krypton point release as well.
@janbar - have you come across this problem yourself? If so, what do you think of my proposed solution.

**Comment by ksooo** [comment]
Yes, your patch ensures that it cannot happen anymore that the timer type of the current timer tag is omitted from the timer types list, but could you please explain how it can happen that a tag is given to the dialog that leads to this problem?

**Comment by metaron-uk** [comment]
* if (type->ForbidsNewInstances()) because only the backend has the necessary UI to create timers of this type, but Kodi can still tweak an existing one (as in the screeshot  example)
* if for some other reason the conditions to create a new rule timer of this type aren't met (although you seem to have fixed a lot of these with your latest PR :-)

**Comment by ksooo** [comment]
The correct fix for the "forbidsnewinstances" case would be to add  "m_bIsNewTimer &&" in front of the respective if clause in the timer types init member function. Seems I forgot this with my latest PR.

Are there any other real cases? Then we should fix those one by one.

Adding the "golden hammer" like you suggest only hides the actual bugs leading to filtering out a type, imo.

**Comment by metaron-uk** [comment]
Adding "m_bIsNewTimer &&" in front covers the 'new timer' case, but what about modifying an existing timer rule of another type?
If the client 'ForbidsNewInstances()' of that timer type, then surely it shouldn't be available for selection as an alternative.
In the same way, if type->RequiresEpgTagOnCreate() && !m_timerInfoTag->GetEpgInfoTag() etc... those wouldn't be valid options to select either.
imo, the only time that an 'invalid' type should appear in the list is when the timer already has that type (aka the "golden hammer").

**Comment by ksooo** [comment]
So, basically you are saying to revert the other "m_bIsNewTimer &&" changes a did in my latest PR and to use the "golden hammer" instead?

**Comment by metaron-uk** [comment]
That would be my recommendation, yes.

**Comment by janbar** [comment]
@metaron-uk , I thought "unhandled" was treated as presented here, regarding its declaration:

    PVR_TIMER_TYPE_IS_REPEATING |
    PVR_TIMER_TYPE_FORBIDS_NEW_INSTANCES |
    PVR_TIMER_TYPE_SUPPORTS_PRIORITY |
    PVR_TIMER_TYPE_SUPPORTS_LIFETIME |
    PVR_TIMER_TYPE_SUPPORTS_RECORDING_GROUP

Seems that isn't the case. I have never been able to test it. But your first screenshot "After (How this type should be presented)" is right for me and this is what I expected.

**Comment by ksooo** [comment]
@metaron-uk you are right. Could you please change this PR to revert my last changes and to implement the "golden hammer"?

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp [hunk 1]
```diff
@@ -769,8 +770,13 @@ void CGUIDialogPVRTimerSettings::InitializeTypesList()
     if (!type->IsTimerRule() && m_timerInfoTag->EndAsLocalTime() < CDateTime::GetCurrentDateTime())
       continue;
 
+    if (type == m_timerType)
```
Comment (ksooo) [inline xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp]: Please change to <code>if (!foundThisType && *type == *m_timertype)</code>.

First is for efficiency, to only compare types if actually needed.
Second is to not compare the timer type pointers but the actual objects.

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp [hunk 2]
```diff
@@ -735,7 +735,8 @@ void CGUIDialogPVRTimerSettings::InitializeTypesList()
     return;
   }
 
-  int idx = 0;
+  bool foundThisType(false);
```
Comment (ksooo) [inline xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp]: I would prefer if you use a type prfix for the variable name => bFoundThisType
We do not do this in a consistent way in our code base, but to follow our coding guidelines we should obey the guideline for new code.

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp [hunk 3]
```diff
@@ -769,8 +770,13 @@ void CGUIDialogPVRTimerSettings::InitializeTypesList()
     if (!type->IsTimerRule() && m_timerInfoTag->EndAsLocalTime() < CDateTime::GetCurrentDateTime())
       continue;
 
+    if (type == m_timerType)
+      foundThisType = true;
+
     m_typeEntries.insert(std::make_pair(idx++, type));
   }
+  if (!foundThisType)
```
Comment (ksooo) [inline xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp]: Pleaes insert a blank line above this line.

**Review by ksooo — CHANGES_REQUESTED** [review]

**Comment by metaron-uk** [comment]
> @metaron-uk , I thought "unhandled" was treated as presented here, regarding its declaration:
> 
> PVR_TIMER_TYPE_IS_REPEATING |
> PVR_TIMER_TYPE_FORBIDS_NEW_INSTANCES |
> PVR_TIMER_TYPE_SUPPORTS_PRIORITY |
> PVR_TIMER_TYPE_SUPPORTS_LIFETIME |
> PVR_TIMER_TYPE_SUPPORTS_RECORDING_GROUP
>
> Seems that isn't the case. I have never been able to test it. But your first screenshot "After (How this type should be presented)" is right for me and this is what I expected.

@janbar You are right, I have been experimenting...

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp (hunk 2 above)
Comment (metaron-uk) [inline xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp]: OK

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp (hunk 1 above)
Comment (metaron-uk) [inline xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp]: Good catch.

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp (hunk 3 above)
Comment (metaron-uk) [inline xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp]: done

**Comment by metaron-uk** [comment]
@ksooo not sure why github is showing both the commits in your last pull request - I only reverted one of them. Let me know if you want this squashed into one.

**Comment by ksooo** [comment]
Please rebase to latest master and then squash the changes of the PR to a single commit.

**Comment by metaron-uk** [comment]
@ksooo - rebased as requested and checked on x86 linux - good to go.

**Comment by ksooo** [comment]
jenkins build this please

**Review by ksooo — APPROVED** [review]
thanks

**Pull request merged**